frida_tools/tracer.py

```diff
@@ -573,6 +603,21 @@ def stop(self):
             self._script = None
 
     def _create_trace_script(self):
+        # Extract all init session files into a single string,
+        # substitute newlines with '\u000a'
+        self._init_session_code = ''
+        for path in self._init_session_paths:
+            # Append contents of <path> to <self._init_session_code>
+            fo = open(path, "rt")
+            line = fo.readlines()
+            fo.close()
+
+            for i in range (len(line)):
+                line [i] = line [i].rstrip ()
+
+            self._init_session_code += '\\u000a'.join(line)
+            self._init_session_code += '\\u000a'
```

Comment:
Wouldn't `\n` have the same effect? Also it would be good to use double-quoted strings here, as these are not enum values.

Reply:
There is a reason for this madness. :-)

I discovered a very important, albeit troubling, behaviour of JavaScript's `eval` function: newlines will cause the eval() call to fail!  This includes two forms of newlines:

(a) **The occurrence of the two adjacent characters backslash ('\') and 'n'**.  Upon reading these within a string, eval() assumes we have come to the end of the line, so the string it was parsing will be unterminated.

(b) **The occurrence of the binary value 0x0A in the input**.  Typically this is due to the developer having pressed ENTER between code lines.  Encountering this causes eval() to throw an error message.

It took me a while to pin this down and solve it.  The solution to both newline problems is:

(a) **The developer must never insert '\n' within the evaluated JS code**.  Instead, s/he should use the String.fromCharCode(10) function call, concatenating this string wherever '\n' is needed.  Although this is equivalent to '\n' at run-time, eval() does not see the '\n' sequence at all.

(b) **Replace the 0x0A between code lines with '\u000a'**.  It turns out eval() accepts this as a valid line delimiter, and does **not** throw an error.  The developer does not have to worry about this -- tracer.py does it for him automatically.

It's interesting to note that *all* eval() examples on the web are toy examples: they always consist of one or more JS statements *on a single line*.  Now we know why they shy away from multiple lines -- because of newlines.  Fortunately, frida-trace can benefit from the workaround.

Reply:
Ahh! It would definitely be good to add a comment explaining this. :-)